docs/development/architecture.org

```diff
@@ -0,0 +1,114 @@
+Nix-Gui may be considered to have two components: the backend and the frontend.
+
+The backend (=nixui/options=) reads information from both NixOS and the configuration path being edited, then converts them to data-structures useful for navigation, search, and editing. The backend maintains in-memory state for all changes and commits those changes to disk by writing to the configuration path.
+
+The frontend (=nixui/graphics=) constructs various widgets comprising Nix-Gui and loads them with data from the backend. The frontend doesn't manage any edit state. All graphical "edits" send an update request to the backend.
+
+* Backend
+The backend handles three major tasks: configuration retrieval to initialize data-structures, in-memory configuration mutation, and configuration commitment. It provides APIs for frontend interaction, but doesn't concern itself with how the frontend renders any of its data.
+
+** Loading the Option Tree Data-Structure
+Nix-Gui queries information from the configuration path and =<nixpkgs>= in order to construct a data-structure (=nixui.options.option_tree.OptionTree=) mapping options and attributes to metadata (=nixui.options.option_tree.OptionData=) including their =NixType=, their =OptionDefinition='s:
+- =_type=: A static python-object representation of the options type, informs the frontend of how =FieldWidget= should be generated and input checking should be performed.
+- =system_default_definition=: Provided by =<nixpkgs>=, the default value if nothing is set in the configuration path.
+- =configured_definition=: Provided by the configuration path, overrides =system_default_definition=.
+- =in_memory_definition=: Undefined when the data-structure is loaded, can be mutated and later committed to disk.
+
+=nixui/nix/lib.nix= (invoked by =nixui.options.nix_eval=) contains the function =get_all_nixos_options()= which is used to retrieve metadata to retrieve =OptionData= for the universe of NixOS options.
```

Comment:
Maybe instead of 'the universe of' say 'all of the', or something like that. Universe feels a bit weird here.

Reply:
for all options supported by NixOS.

for all options allowed by the NixOS config schema.

("schema" is the technical term for this thing)

Reply:
resolved with https://github.com/nix-gui/nix-gui/pull/214/commits/6baaca2fce50ccc18d80b6674a4db092ff1df1a1